internal/codeintel/bundles/persistence/cache/reader_cache.go

```diff
@@ -0,0 +1,251 @@
+	c.readers[key] = entry
+
+	go func() {
+		// Initialize the reader in the background. This allows multiple concurrent requests
+		// for the same reader to share the same entry while it is initializing, which an take
+		// a non-negligible amount of time for large, outdated bundles.
+		entry.init(func() (persistence.Reader, error) { return c.opener(key) })
+
+		// Block until the entry is unused, then close the underlying reader
+		entry.closeOnceUnused()
+
+		// Remove entry from map so that it's inaccessible to future requests
+		c.removeEntry(key)
+	}()
+
+	return entry, true
+}
+
+// removeEntry removes the given key from the cache's reader map.
+func (c *readerCache) removeEntry(key string) {
+	c.m.Lock()
+	delete(c.readers, key)
+	c.m.Unlock()
+}
+
+// makeInitChannel creates a channel that closes once the given entry has initialized.
+func makeInitChannel(entry *readerCacheEntry) <-chan struct{} {
+	ch := make(chan struct{})
+
+	go func() {
+		defer close(ch)
+		entry.waitUntilInitialized()
+	}()
+
+	return ch
+}
+
+// readerCacheEntry wraps a reader with its initialization state and its ref count.
+type readerCacheEntry struct {
+	m           *sync.Mutex        // protects all fields
+	reader      persistence.Reader // shared reader instance
+	err         error              // construction error
+	initialized bool               // set when reader/err fields are set
+	disposed    bool               // set when entry is no longer usable
+	refCount    uint32             // number of references to entry
+
+	// cond wraps the entry's mutex and broadcasts to waiting
+	// goroutines when initialized is first set to true and when
+	// refCount is decremented.
+	cond *sync.Cond
+}
+
+// newReaderCacheEntry creates a new cache entry with a refcount of one.
+func newReaderCacheEntry() *readerCacheEntry {
+	m := &sync.Mutex{}
+
+	return &readerCacheEntry{
+		m:        m,
+		cond:     sync.NewCond(m),
+		refCount: 1,
+	}
+}
+
+// init calls the given function to construct a reader, then updates the reader, err,
+// and initialized fields of the entry.
+func (e *readerCacheEntry) init(openReaderFunc func() (persistence.Reader, error)) {
+	reader, err := openReaderFunc()
+
+	e.m.Lock()
+	e.reader = reader
+	e.err = err
+	e.initialized = true
+	e.m.Unlock()
+
+	// wake users waiting for initialization to complete
+	e.cond.Broadcast()
+}
+
+// waitUntilInitialized blocks until the entry has completed initialization.
+func (e *readerCacheEntry) waitUntilInitialized() {
+	e.m.Lock()
+	defer e.m.Unlock()
+
+	for !e.initialized {
+		e.cond.Wait()
+	}
+}
+
+// closeOnceUnused blocks until the refcount of the entry goes to zero. The entry's disposed
+// flag is set to indicate to future users that the underlying reader is no longer usable. The
+// entry's reader is then closedK (if it was initialized successfully).
+func (e *readerCacheEntry) closeOnceUnused() {
+	e.m.Lock()
+	for e.refCount > 0 {
+		e.cond.Wait()
+	}
+	e.disposed = true
+	e.m.Unlock()
```

Comment:
It is very unlikely but still possible case I think would cause deadlock: `closeOnceUnused()` is called before `waitUntilInitialized()`. They both acquire `e.m` but `closeOnceUnused()` will not release the mutex until `e.refCount` becomes 0, which makes the cache entry only "mark as initialized" when no users to it (including the very first user who initiated creation of the cache entry).

However, the real deadlock is mitigated by checking context deadline in `WithReader` method. Therefore totally it's up to you.

Here is my suggested change:

```suggestion
	for atomic.LoadUint32(e.refCount) > 0 {
		e.cond.Wait()
	}
	e.m.Lock()
	e.disposed = true
	e.m.Unlock()
```

To completely get rid of using mutex in this and some other methods, you can also change `disposed` to be `uint32` as well. Then use `0/1` indicates `false/true` and [`atomic` package](https://golang.org/pkg/sync/atomic) to manipulate.

Reply:
> closeOnceUnused() will not release the mutex until e.refCount becomes 0

condition variables are tricky :) `Wait()` releases the mutex on calling, then acquires again once it is signaled. So this isn't an issue. We don't need to use atomics because all reads/writes are protected by the mutex. In fact using atomics leads to issues because there are changes here that happen across fields, but they must happen atomically. Hence the switch to the use of a single mutex and a condition variable.

Reply:
`e.cond.Wait()` would release `e.m` on calling? 🤔 How do they get connected...